src/interfaces/records/handlers/records-upload.ts

```diff
@@ -0,0 +1,196 @@
+
+    try {
+      const existingUploadComplete = await RecordsUpload.getUploadComplete(existingMessages);
+      if (existingUploadComplete) {
+        return new MessageReply({
+          status: { code: 400, detail: 'upload already complete' }
+        });
+      }
+    } catch (e) {
+      // ignore error if upload complete cannot be found
+    }
+
+    const newMessageIsUploadStart = await recordsUpload.isUploadStart();
+    const newMessageIsUploadComplete = await recordsUpload.isUploadComplete();
+
+    const parts = new Set<number>();
+
+    try {
+      const existingUploadStart = await RecordsUpload.getUploadStart(existingMessages);
+      if (existingUploadStart) {
+        if (newMessageIsUploadStart) {
+          return new MessageReply({
+            status: { code: 400, detail: 'cannot start an upload more than once' }
+          });
+        }
+
+        RecordsUpload.verifyEqualityOfImmutableProperties(existingUploadStart, incomingMessage);
+      }
+
+      for (const existingMessage of existingMessages) {
+        RecordsUpload.verifyExclusivityOfUniqueProperties(existingMessage, incomingMessage);
+
+        if (newMessageIsUploadComplete) {
+          if (existingMessage.descriptor.state === DwnStateName.Part) {
+            const uploadPart = existingMessage as RecordsUploadPartMessage;
+
+            parts.add(uploadPart.descriptor.index);
+          }
+        }
+      }
+    } catch (e) {
+      return new MessageReply({
+        status: { code: 400, detail: e.message }
+      });
+    }
+
+    if (newMessageIsUploadComplete) {
+      const uploadComplete = incomingMessage as RecordsUploadCompleteMessage;
+
+      for (let index = 0; index < uploadComplete.descriptor.count; ++index) {
+        if (!parts.has(index)) {
+          return new MessageReply({
+            status: { code: 400, detail: `missing index '${index}'` }
+          });
+        }
+
+        parts.delete(index);
+      }
+
+      for (const index of parts) {
+        return new MessageReply({
+          status: { code: 400, detail: `extra index '${index}'` }
+        });
+      }
+    }
+
+    let result;
+
+    switch (incomingMessage.descriptor.state) {
+    case DwnStateName.Complete:
+      var uploadComplete = incomingMessage as RecordsUploadCompleteMessage;
+      result = await this.uploadStore.complete(tenant, uploadComplete.recordId, uploadComplete.descriptor.count);
+      break;
+
+    case DwnStateName.Part:
+      var uploadPart = incomingMessage as RecordsUploadPartMessage;
+      result = await this.uploadStore.part(tenant, uploadPart.recordId, uploadPart.descriptor.index, dataStream);
+      break;
+
+    case DwnStateName.Start:
+      var uploadStart = incomingMessage as RecordsUploadStartMessage;
+      result = await this.uploadStore.start(tenant, uploadStart.recordId, uploadStart.descriptor.dataFormat);
+      break;
+    }
+
+    if (!result) {
```

Comment:
Could this go as a `default` case in the `switch` statement?

Reply:
no.  `result` is checking that the `UploadStore` operation did what was expected (e.g. the `dataCid` and `dataSize` match), not that we're handling some unknown `DwnStateName` value